test/apps/iodemo/ucx_wrapper.cc

```diff
@@ -0,0 +1,575 @@
+{
+    ucp_request_cancel(_worker, _iomsg_recv_request);
+    try {
+        request_wait("iomsg_recv", _iomsg_recv_request);
+    } catch (const UcxError& e) {
+        log() << "io msg request is completed with status " << e.status()
+              << std::endl;
+    }
+
+    ucp_worker_destroy(_worker);
+}
+
+UcxContext::~UcxContext()
+{
+    cleanup_conns();
+    cleanup_listener();
+    cleanup_worker();
+    ucp_cleanup(_context);
+}
+
+void UcxContext::listen(const struct sockaddr* saddr, size_t addrlen)
+{
+    ucp_listener_params_t listener_params;
+
+    listener_params.field_mask         = UCP_LISTENER_PARAM_FIELD_SOCK_ADDR |
+                                         UCP_LISTENER_PARAM_FIELD_CONN_HANDLER;
+    listener_params.sockaddr.addr      = saddr;
+    listener_params.sockaddr.addrlen   = addrlen;
+    listener_params.conn_handler.cb    = connect_callback;
+    listener_params.conn_handler.arg   = reinterpret_cast<void*>(this);
+
+    ucs_status_t status = ucp_listener_create(_worker, &listener_params,
+                                              &_listener);
+    if (status != UCS_OK) {
+        throw UcxError("ucp_listener_create", status);
+    }
+
+    log() << "started listener " << _listener
+          << " on " << sockaddr_str(saddr, addrlen) << std::endl;
+}
+
+UcxConnection* UcxContext::connect(const struct sockaddr* saddr, size_t addrlen)
+{
+    UcxConnection *conn = new UcxConnection(*this, get_next_conn_id(), saddr,
+                                            addrlen);
+    add_connection(conn);
+    return conn;
+}
+
+void UcxContext::disconnect(UcxConnection* conn)
+{
+    verbose_os() << "closing connection " << conn << " with id = " << conn->id()
+                 << std::endl;
+    remove_connection(conn);
+    if (!conn->disconnect()) {
+        _closing_conns.push_back(conn);
+    }
+}
+
+void UcxContext::on_disconnect(UcxConnection* conn) _GLIBCXX_NOTHROW
+{
+    verbose_os() << "closing connection " << conn << " with id = " << conn->id()
+                 << std::endl;
+    remove_connection(conn);
+    if (!conn->on_disconnect()) {
+        _closing_conns.push_back(conn);
+    }
+}
+
+void UcxContext::dispatch_new_connection(UcxConnection *conn)
+{
+}
+
+void UcxContext::dispatch_io_message(UcxConnection* conn, const void *buffer)
+{
+}
+
+void UcxContext::progress()
+{
+    ucp_worker_progress(_worker);
+
+    if (_iomsg_recv_request->completed) {
+        process_io_message();
+        request_release(_iomsg_recv_request);
+        post_recv();
+    }
+
+    process_conn_request();
+
+    if (!_closing_conns.empty()) {
+        UcxConnection *conn = _closing_conns.front();
+        if (conn->disconnect()) {
+            _closing_conns.pop_front();
+            delete conn;
+        }
+    }
+}
+
+void UcxContext::request_wait(const char *what, void *request)
+{
+    if (UCS_PTR_IS_PTR(request)) {
+        ucs_status_t status;
+        do {
+            ucp_worker_progress(_worker);
```

Comment:
need to progress whole context

Reply:
this may cause stack overflow on closing connections flow